feat: k6 generator

jakowenko opened the pull request.

jakowenko committed:
feat: k6 generator

jakowenko commented on packages/k6-generator/src/unique/user.json:
```diff
@@ -0,0 +1,16 @@
+{
```
This file before was too large. Almost 15MB. If we need one with that many records, can we dynamically make them? Why do they need to be in a static file.

roselandroche replied:
Dynamically creating the information was an issue I ran into due to the nature of k6 and imports, but I can revisit this. 

Now that I've nailed the deleting/logging out during the actual test though, the size of this file can be drastically reduced with no impact on the test - I just have to rerun the model generator for a smaller output

jakowenko commented on packages/k6-generator/src/index.js:
```diff
@@ -0,0 +1,260 @@
+  const methods = require(`${__dirname}${methodsPath.slice(methodsPath.indexOf('/'))}`);
+
+  const newFile = await new StringPrompt({
+    initial: 'tests/k6/scripts/alpha.js',
```
Should we default this to something from the input file? If the input file is `expressTemplate.json`, we can default the filename to something like `expressTemplate.k6.js`

jakowenko commented on packages/k6-generator/src/index.js:
```diff
@@ -0,0 +1,260 @@
+    let methodUrl = Object.keys(methods.routes)[i];
+
+    for (let j = 0; j < methods.routes[methodUrl].length; j++) {
+      const current = methods.routes[methodUrl][j];
```
This and other places you can just do:

```const { path, uniquePayload } = methods.routes[methodUrl][j];```

jakowenko commented on packages/k6-generator/src/index.js:
```diff
@@ -0,0 +1,260 @@
+  }
+
+  if (!uniqueObj) {
+    uniqueObj = '""';
```
What is this for?

roselandroche replied:
This is because replace-in-file doesn't keep the first set of quotes - so when there was no unique object to define, I would end up with something like `uniqueObj = ;` which would give me an error

jakowenko commented on packages/k6-generator/src/index.js:
```diff
@@ -0,0 +1,260 @@
+    let methodUrl = Object.keys(methods.routes)[i];
+
+    for (let j = 0; j < methods.routes[methodUrl].length; j++) {
+      const current = methods.routes[methodUrl][j];
```
```const { tag } = methods.routes[methodUrl][j];```

jakowenko commented on packages/k6-generator/src/index.js:
```diff
@@ -0,0 +1,260 @@
+  writeVuObj = `let vuObj = ${JSON.stringify(vuObj)};\n\n`;
+
+  // PARSE THROUGH ALL ROUTES AND ADD TO MAIN CONTENT - WITHIN DEFAULT FUNCTION
+  const baseUrl = methods['baseUrl'];
```
```const { baseUrl } = methods;```

jakowenko commented on packages/k6-generator/src/index.js:
```diff
@@ -0,0 +1,260 @@
+    let fullPath;
+
+    for (let j = 0; j < methods.routes[methodUrl].length; j++) {
+      let current = methods.routes[methodUrl][j];
```
Same as above.

jakowenko commented on packages/k6-generator/src/index.js:
```diff
@@ -0,0 +1,260 @@
+        customImport = `const SLEEP_DURATION = ${methods['sleep']}\nconst uniqueData = JSON.parse(open('${uniqueFile}'))`;
+        uniqueObj = `'${uniqueFile.slice(uniqueFile.lastIndexOf('/') + 1, -5)}'`;
+      }
+      if (!customImport) {
```
I would do something like this for `customImport`.

```
customImport = uniquePayload ? `const SLEEP_DURATION = ${methods['sleep']}\nconst uniqueData = JSON.parse(open('${uniqueFile}')) : `const SLEEP_DURATION = ${methods['sleep']}`
```

Also, I would try to move the `const SLEEP_DURATION` to the template. And only set the value of it from this generate file.

roselandroche replied:
I couldn't do the exact code example you gave for custom import because of where it was in the code, but I did refactor to something that looks cleaner! And sleep duration is also a separate prompt to replace a value in the template now

jakowenko commented on packages/k6-generator/src/index.js:
```diff
@@ -0,0 +1,260 @@
+        payload = [];
+        tagName = '${tag}Res';`;
+
+      if (custom) {
```
What does custom do?

roselandroche replied:
The `custom` property indicates if there is something dynamic in the path, for example a `delete` path requiring an `${id}`. I parse the full path differently in this case

jakowenko commented on packages/k6-generator/src/index.js:
```diff
@@ -0,0 +1,260 @@
+        case 'createUser':
+        case 'login':
+        case 'logout':
+        case 'post':
```
If you want to get fancy, a lot of this can probably be put into a method template file then find and replace similar to what you are doing with the base k6 one.

roselandroche committed:
wip: k6 generator cleanup

- prompt for sleep
- initial new file name upgraded
- more unique data to use

roselandroche committed:
wip: k6 generator cleanup

- generated report has dynamic name and path
- docs update
- general clean up

roselandroche committed:
wip: refactor

- routes can go in any logical order
- summary prompt upgrade

roselandroche committed:
fix: intentional auth

- auth added on login and set to null on logout

roselandroche committed:
wip: refactor and clean

- no more floating `false;`
- setNull property to remove id and token
- docs update

roselandroche committed:
docs: minor edits

roselandroche committed:
wip: k6 generator tweak

- less unique data needed
- now we'll need as many unique items as there are simultaneous VU's

roselandroche committed:
fix: vu id in test

- update to account for zero indexing

roselandroche committed:
refactor: based on k6 warning

- https://k6.io/docs/examples/data-parameterization/
- use of Shared array to diminish memory issue

roselandroche committed:
refactor: naming, location

- moved unique/user.json and adjusted checks
- renamed summary

roselandroche committed:
wip: resolve pathing

- use path.resolve to avoid relative pathing issues
- rename 'path' in methods file to fix naming overlap
